plugins/repository-s3/qa/amazon-s3/src/test/java/org/elasticsearch/repositories/s3/AmazonS3TestServer.java

```diff
@@ -106,13 +106,38 @@ public Response handle(final String method,
         }
 
         final List<String> authorizations = headers.get("Authorization");
-        if (authorizations == null
-            || (authorizations.isEmpty() == false & authorizations.get(0).contains("s3_integration_test_access_key") == false)) {
-            return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "Access Denied", "");
+        if (authorizations == null || authorizations.size() != 1) {
+            return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "No authorization", "");
+        }
+        final String authorization = authorizations.get(0);
+        final String permittedBucket;
+        if (authorization.contains("s3_integration_test_permanent_access_key")) {
+            final List<String> sessionTokens = headers.get("x-amz-security-token");
+            if (sessionTokens != null) {
+                return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "Unexpected session token", "");
+            }
+            permittedBucket = "permanent_bucket_test";
+        } else if (authorization.contains("s3_integration_test_temporary_access_key")) {
+            final List<String> sessionTokens = headers.get("x-amz-security-token");
+            if (sessionTokens == null || sessionTokens.size() != 1) {
+                return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "No session token", "");
+            }
+            final String sessionToken = sessionTokens.get(0);
+            if (sessionToken.equals("s3_integration_test_temporary_session_token") == false) {
+                return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "Bad session token", "");
+            }
+            permittedBucket = "temporary_bucket_test";
+        } else {
+            return newError(requestId, RestStatus.FORBIDDEN, "AccessDenied", "Bad access key", "");
         }
 
         final RequestHandler handler = handlers.retrieve(method + " " + path, params);
         if (handler != null) {
+            final String bucket = params.get("bucket");
+            if (bucket != null && permittedBucket.equals(bucket) == false) {
+                // allow a null bucket to support bucket-free APIs like ListBuckets?
```

Comment:
>  allow a null bucket to support bucket-free APIs like ListBuckets?

I don't think it's needed, returning an error seems the right thing to do

Reply:
It is needed, but it's not `ListBuckets` so I pushed 63caae2bd24. It's the multi-object delete API, which puts the bucket name in the `Host` header rather than in a URL parameter, which I do not think is worth adding the extra plumbing for.